hathor/cli/top.py

```diff
@@ -0,0 +1,704 @@
+        """Initialize the screen and immediately draw an initial screen."""
+        self.stdscr = curses.initscr()
+        curses.noecho()
+        curses.cbreak()
+        curses.curs_set(0)
+        self.stdscr.keypad(True)
+        self.stdscr.nodelay(True)
+        self.register_colors()
+
+        self._redraw: bool = False
+
+        self.freeze_data: bool = False
+
+        self.loop: AbstractEventLoop = loop
+        self.loop.add_reader(sys.stdin, self.getch)
+
+        self.fetcher: 'DataFetcher' = fetcher
+
+        self.screen_list: Dict[str, Window] = {
+            'help': HelpWindow(self, self.stdscr),
+            'main': MainWindow(self, self.stdscr),
+            'control': ControlWindow(self, self.stdscr),
+        }
+        self.screen = self.screen_list['main']
+
+        # Capture Ctrl+C.
+        def signal_handler(sig, frame):
+            self.quit()
+        signal.signal(signal.SIGINT, signal_handler)
+
+        # Capture resize.
+        def resize_signal_handler(sig, frame):
+            import fcntl
+            import struct
+            import termios
+            try:
+                h, w = struct.unpack('hh', fcntl.ioctl(1, termios.TIOCGWINSZ, '1234'))
+            except Exception:
+                w = curses.tigetnum('cols')
+                h = curses.tigetnum('lines')
+            curses.resizeterm(h, w)
+            self.redraw()
+        signal.signal(signal.SIGWINCH, resize_signal_handler)
+
+        # Fetch task.
+        self.data: Optional[ProfileData] = None
+        self.fetcher.on_fetch_success = self.on_data_update
+        self.fetcher.on_fetch_error = self.redraw
+        self.fetcher.start()
+
+        # Finally, draw for the first time.
+        self.redraw()
+
+    def register_colors(self):
+        """Register colors used by the windows."""
+        global Color
+        curses.start_color()
+        Color = DefaultColor()
+
+    def on_data_update(self, data: ProfileData) -> None:
+        if self.freeze_data:
+            return
+        self.data = data
+        self.screen.on_data_update()
+        self.redraw()
+
+    def goto(self, name):
+        self.screen = self.screen_list[name]
+        self.redraw()
+
+    def redraw(self):
+        if self._redraw:
+            return
+        self._redraw = True
+        self.loop.call_later(0, self.render)
+
+    def getch(self):
+        key = self.stdscr.getch()
+        self.on_keypress(key)
+
+    def on_keypress(self, key):
+        self.screen.on_keypress(key)
+
+    def render(self):
+        self.stdscr.clear()
+        try:
+            self.screen.render()
+        except Exception as e:
+            self.quit()
+            raise e
+        self.stdscr.refresh()
+        self._redraw = False
+
+    def quit(self):
+        curses.nocbreak()
+        self.stdscr.keypad(False)
+        curses.echo()
+        curses.endwin()
+        self.loop.stop()
+
+
+class DataFetcher:
```

Comment:
Missing type annotations on nearly all methods.

Reply:
Done. And I also renamed the class to `ProfilerAPIClient`.